src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10949,99 +10946,143 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
             // parse out a missing name node for the expression, and keep on going
             var tk = this.CurrentToken.Kind;
             if (IsInvalidSubExpression(tk))
-            {
                 return this.AddError(this.CreateMissingIdentifierName(), ErrorCode.ERR_InvalidExprTerm, SyntaxFacts.GetText(tk));
-            }
 
-            // Parse a left operand -- possibly preceded by a unary operator.
-            if (IsExpectedPrefixUnaryOperator(tk))
-            {
-                var opKind = SyntaxFacts.GetPrefixUnaryExpression(tk);
-                newPrecedence = GetPrecedence(opKind);
-                var opToken = this.EatToken();
-                var operand = this.ParseSubExpression(newPrecedence);
-                leftOperand = _syntaxFactory.PrefixUnaryExpression(opKind, opToken, operand);
-            }
-            else if (tk == SyntaxKind.DotDotToken)
+            return ParseExpressionContinued(parseUnaryOrPrimaryExpression(precedence), precedence);
```

Comment:
extracted out common patters to make things much clearer.  The core algorithm is now simple to reason about.  We first parse out a unary/primary expression.  Then, we parse out what can continue that at the current precedence we're at.